### PR TITLE
[SPARK-47141][CORE] Support enabling migration of shuffle data directly to external storage using config parameter.

### PR DESCRIPTION
**Pull request by maheshk114**
### What changes were proposed in this pull request?
This PR exposes a config allowing user to migrate the shuffle data directly to external storage. Changes are made to migrate the data in multiple thread to reduce the migration time. Similarly, the reading of shuffle data from external storage is done using multiple thread. Configuration parameter is added to control the number of threads to be used for reading shuffle data.

### Why are the changes needed?
Currently Spark supports migration of shuffle data to peer nodes during node decommissioning. If peer nodes are not accessible, then Spark falls back to external storage. User needs to provide the storage location path. There are scenarios where user may want to migrate to external storage instead of peer nodes. This may be because of unstable nodes or due to the need of aggressive scale down. So, user should be able to configure to migrate the shuffle data directly to external storage if the use case permits. 

### Does this PR introduce _any_ user-facing change?

Yes, now user can enable migrating shuffle data directly to external storage and also control the number of threads used for reading shuffle data from external storage.

### How was this patch tested?

New Unit tests are added.
Made sure that existing tests are passing.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by mridulm** [comment]
+CC @dongjoon-hyun , @Ngone51

**Comment by abhishekd0907** [comment]
Hi @mridulm  @attilapiros , Can you please help in reviewing this PR?

**Comment by mridulm** [comment]
+CC @dongjoon-hyun, @Ngone51

**Comment by dongjoon-hyun** [comment]
Sorry for missing pings here, @mridulm . Let me take a look during this week.

FYI, I was thinking that Apache Hadoop 3.4 and S3 Express One Zone feature was the minimum limitation of this efforts because historically these kind of previous effort didn't deliver desirable benefits due to the inevitable slowness.

I'll try quickly when I have some time and share the opinion here.

**Comment by dongjoon-hyun** [comment]
I assigned myself not to forget.

**Comment by mridulm** [comment]
Thanks @dongjoon-hyun :-)
 (And I forgot I had pinged you before, sorry about that !)

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
To @maheshk114 , I have a big concern on the AS-IS PR title because it is misleading the users by ignoring the existing Apache Spark 3.2 features.

> [SPARK-47141] [Core]: Support shuffle migration to external storage.

At least, as the author of the following Apache Spark 3.2 patches, I believe Apache Spark "Support Shuffle Migration to External Storage" already. WDTY, @maheshk114 ? Could you revise the PR title to narrow down to this PR's exact contribution?
- #30492
- #31215

**Comment by dongjoon-hyun** [comment]
To @maheshk114 , could you rebase this PR to the `master` branch once more to test on top of Apache Hadoop 3.4.0? The following is merged to Apache Spark `master` branch.
- #45583

**Comment by maheshk114** [comment]
> To @maheshk114 , I have a big concern on the AS-IS PR title because it is misleading the users by ignoring the existing Apache Spark 3.2 features.
> 
> > [SPARK-47141] [Core]: Support shuffle migration to external storage.
> 
> At least, as the author of the following Apache Spark 3.2 patches, I believe Apache Spark "Support Shuffle Migration to External Storage" already. WDTY, @maheshk114 ? Could you revise the PR title to narrow down to this PR's exact contribution?
> 
> * [[SPARK-33545][CORE] Support Fallback Storage during Worker decommission #30492](https://github.com/apache/spark/pull/30492)
> * [[SPARK-34142][CORE] Support Fallback Storage Cleanup during stopping SparkContext #31215](https://github.com/apache/spark/pull/31215)

Thanks a lot @dongjoon-hyun for looking into the PR. Yes, it make sense. I have updated the PR title to reflect the exact changes done in this PR.

**Comment by maheshk114** [comment]
> To @maheshk114 , could you rebase this PR to the `master` branch once more to test on top of Apache Hadoop 3.4.0? The following is merged to Apache Spark `master` branch.
> 
> * [[SPARK-45393][BUILD] Upgrade Hadoop to 3.4.0 #45583](https://github.com/apache/spark/pull/45583)

done

**Comment by dongjoon-hyun** [comment]
Thank you for rebasing, @maheshk114 .

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala [hunk 1]
```diff
@@ -447,7 +462,30 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
...
+    indexOut.writeLong(10L) // next offset
+    indexOut.writeLong(10L) // next offset
+    val dataOut = fallbackFileSystem.create(dataFile)
+    dataOut.writeBytes("some data to write")
+    indexOut.close()
+    dataOut.close()
+  }
+
+  test("fetch continuous blocks in batch successful 3 local + 4 host local + 2 remote reads " +
+    "+ 2 read from external storage") {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala:488]: This PR seems to introduce a test failure (or flakiness) at this test case. Could you take a look at the CI failure?
- https://github.com/maheshk114/spark/actions/runs/8369424178/job/22915159186
```
[info] ShuffleBlockFetcherIteratorSuite:
...
[info] - fetch continuous blocks in batch successful 3 local + 4 host local + 2 remote reads + 2 read from external storage *** FAILED *** (107 milliseconds)
[info]   org.mockito.exceptions.verification.TooFewActualInvocations: blockManager.getLocalBlockData(<any>);
[info] Wanted 2 times:
[info] -> at org.apache.spark.storage.BlockManager.getLocalBlockData(BlockManager.scala:721)
[info] But was 1 time:
[info] -> at org.apache.spark.storage.ShuffleBlockFetcherIterator.fetchLocalBlocks(ShuffleBlockFetcherIterator.scala:592)
```

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 2]
```diff
@@ -551,6 +551,22 @@ package object config {
       .checkValue(_.endsWith(java.io.File.separator), "Path should end with separator.")
       .createOptional
 
+  private[spark] val STORAGE_FALLBACK_STORAGE_NUM_THREADS_FOR_SHUFFLE_READ =
+    ConfigBuilder("spark.storage.fallbackStorage.num.threads.for.shuffle.read")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:555]: Every `.` means a new namespace in Apache Spark configuration namespace scheme. For example, this line introduces 4 new namespaces. Please avoid using `.`.
- spark.storage.fallbackStorage.num.*
- spark.storage.fallbackStorage.num.threads.*
- spark.storage.fallbackStorage.num.threads.for.*
- spark.storage.fallbackStorage.num.threads.for.shuffle.*

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala [hunk 3]
```diff
@@ -354,7 +355,8 @@ private[spark] class CoarseGrainedExecutorBackend(
                 // We can only trust allBlocksMigrated boolean value if there were no tasks running
                 // since the start of computing it.
                 if (allBlocksMigrated && (migrationTime > lastTaskFinishTime.get())) {
-                  logInfo("No running tasks, all blocks migrated, stopping.")
+                  val timeTakenMs = (System.nanoTime() - startTime) / (1000 * 1000)
+                  logInfo(s"No running tasks, all blocks migrated in $timeTakenMs ms, stopping.")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:359]: Please revert the log-related change of this file because it's irrelevant to the functionality. We can handle this separately if needed.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 4]
```diff
@@ -92,6 +95,57 @@ private[storage] class FallbackStorage(conf: SparkConf) extends Logging {
...
+          })
+        )
+      case _ =>
+        logInfo(s" fetchThreadPool does not exists for $address or shutdown")
+        blocks.foreach(block => fetchShuffleBlocks(block, blockManager, listener))
+    }
+  }
+
+  private def fetchShuffleBlocks(
+              block: FetchBlockInfo,
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:132]: indentation.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 5]
```diff
@@ -147,6 +222,14 @@ private[spark] object FallbackStorage extends Logging {
     }
   }
 
+  def stopThreadPool(conf: SparkConf): Unit = {
+    logInfo(s" Stopping thread pool")
+    if (getFallbackStorage(conf).isDefined &&
+      getFallbackStorage(conf).get.fetchThreadPool.isDefined) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:228]: indentation.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 6]
```diff
@@ -188,15 +271,15 @@ private[spark] object FallbackStorage extends Logging {
...
         val startTimeNs = System.nanoTime()
-        f.seek(offset)
-        f.readFully(array)
-        logDebug(s"Took ${(System.nanoTime() - startTimeNs) / (1000 * 1000)}ms")
-        f.close()
+        Utils.tryWithResource(fallbackFileSystem.open(dataFile)) { f =>
+          f.seek(offset)
+          f.readFully(array)
+          logDebug(s"Took ${(System.nanoTime() - startTimeNs) / (1000 * 1000)}ms")
+        }
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:282]: Could you spin off this `Utils.tryWithResource` related change to a new JIRA? We can merge that first.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 7]
```diff
@@ -1558,6 +1587,7 @@ object ShuffleBlockFetcherIterator {
       mapIndex: Int,
       address: BlockManagerId,
       size: Long,
+      timeTaken: Long,
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:1590]: May I ask the reason why we add a new filed here? If this is only for pure logging, please revert this.

File: core/src/test/scala/org/apache/spark/storage/FallbackStorageSuite.scala [hunk 8]
```diff
@@ -172,61 +172,64 @@ class FallbackStorageSuite extends SparkFunSuite with LocalSparkContext {
   }
 
   test("migrate shuffle data to fallback storage") {
-    val conf = new SparkConf(false)
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/storage/FallbackStorageSuite.scala:175]: Please preserve the existing test coverage as much as possible. We had better add new feature test coverage separately while keeping the existing test coverage unchanged in order to prevent any regression.

**Review by dongjoon-hyun — COMMENTED** [review]
I finished the first round review, @maheshk114 .

BTW, could you share some performance result of this contribution on your system? Please add them to the PR description.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 6 above)
Comment (maheshk114) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:282]: @dongjoon-hyun  https://github.com/apache/spark/pull/45663

**Review by dongjoon-hyun — COMMENTED** [review]
#45663 is merged. Please rebase this PR to the `master` branch once more, 
@maheshk114 .

**Comment by maheshk114** [comment]
> I finished the first round review, @maheshk114 .
> 
> BTW, could you share some performance result of this contribution on your system? Please add them to the PR description.

I have a internal setup with 3.4 ..will share the number on that.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 9]
```diff
@@ -398,8 +410,13 @@ final class ShuffleBlockFetcherIterator(
     var pushMergedLocalBlockBytes = 0L
     val prevNumBlocksToFetch = numBlocksToFetch
 
-    val fallback = FallbackStorage.FALLBACK_BLOCK_MANAGER_ID.executorId
-    val localExecIds = Set(blockManager.blockManagerId.executorId, fallback)
+    // Fallback to original implementation, if thread pool is not enabled.
+    val localExecIds = if (FallbackStorage.getNumReadThreads(blockManager.conf) > 0) {
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:414]: @maheshk114 can you explain why this change is needed?

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 10]
```diff
@@ -110,14 +158,35 @@ private[spark] object FallbackStorage extends Logging {
   /** We use one block manager id as a place holder. */
   val FALLBACK_BLOCK_MANAGER_ID: BlockManagerId = BlockManagerId("fallback", "remote", 7337)
 
-  def getFallbackStorage(conf: SparkConf): Option[FallbackStorage] = {
-    if (conf.get(STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).isDefined) {
-      Some(new FallbackStorage(conf))
+  // There should be only one fallback storage thread pool per executor.
+  var fallbackStorage: Option[FallbackStorage] = None
+  def getFallbackStorage(conf: SparkConf): Option[FallbackStorage] = this.synchronized {
+    if (conf != null && conf.get(STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).isDefined) {
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:164]: Why the null checks for `conf`?

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 11]
```diff
@@ -110,14 +158,35 @@ private[spark] object FallbackStorage extends Logging {
...
+        fallbackStorage
+      }
     } else {
       None
     }
   }
 
+  def getNumReadThreads(conf: SparkConf): Int = {
+    val numShuffleThreads =
+      if (conf == null) None else conf.get(STORAGE_FALLBACK_STORAGE_NUM_THREADS_FOR_SHUFFLE_READ)
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:186]: Same here:
Why the null checks for `conf`?

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 9 above)
Comment (maheshk114) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:414]: @attilapiros Thanks for taking a look at this PR. In the current code, the shuffle data read from local disk and external storage is handled in same way. It first tries to read from local disk and if it fails then it tries to read from external storage(BlockManager.getLocalBlockData). In this PR I have modified this behavior. The shuffle data written to external storage will be treated as remote block. If the remote block id points to external storage, then the read will be done in a separate thread. This code change is to make sure that, if fallback storage read thread pool is configured, then the fallback storage (external storage) blocks will be processed as remote block or else it will be processed as local block.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 10 above)
Comment (maheshk114) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:164]: Some unit tests scenario, the conf is not initialized.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 11 above)
Comment (maheshk114) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:186]: to fix UT failures.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 10 above)
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:164]: I would prefer to have the fix in the unit test and not in the production code. Otherwise the next developer touching this code will be uncertain why having null checks for the config here and there and he might keep adding null checks to his new code. Moreover if the config would be optional it should be an `Option` but this mixed solution is a bit misleading.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 12]
```diff
@@ -92,6 +95,51 @@ private[storage] class FallbackStorage(conf: SparkConf) extends Logging {
...
+        )
+      case _ =>
+        logInfo(s" fetchThreadPool does not exists for $address or shutdown")
+        blocks.foreach(block => fetchShuffleBlocks(block, blockManager, listener))
+    }
+  }
+
+  private def fetchShuffleBlocks(
+      block: FetchBlockInfo,
+      blockManager: BlockManager,
```
Comment (attilapiros) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:133]: Why `blockManager` is passed if never used?

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 9 above)
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:414]: I see that but what I do not get why we have two separate behaviour at the read side.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 10 above)
Comment (maheshk114) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:164]: make sense. Will modify accordingly.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 9 above)
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:414]: I found where it was first introduced:
https://github.com/apache/spark/pull/30492/files#r532833817

@dongjoon-hyun can an externally stored block ever be found locally?

If not I would prefer to see a separate case for the external blocks and count them separately too.
So we would have the following block types:
- local
- host-local
- push-merged-local
- remote
- external

@maheshk114, @dongjoon-hyun WDYT?

Comment (maheshk114) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:414]: Yes that would be a proper fix.

**Comment by attilapiros** [comment]
I am also concerned about the performance.
 
I think the best would be if the migration of shuffle data to external storage would only kick in when the scale down is aggressive. This can be decided by checking the ratio of the number of available peers (non-decommissioning executors) and the number of decommissioning executors. In that case the parameter would not be a single boolean flag but a threshold for the ratio.

@maheshk114 WDYT?

**Comment by maheshk114** [comment]
> I am also concerned about the performance.
> 
> I think the best would be if the migration of shuffle data to external storage would only kick in when the scale down is aggressive. This can be decided by checking the ratio of the number of available peers (non-decommissioning executors) and the number of decommissioning executors. In that case the parameter would not be a single boolean flag but a threshold for the ratio.
> 
> @maheshk114 WDYT?

Its not only performance but also useful when the nodes are not very reliable. So I think we should have a Boolean flag also to allow user to chose to migrate the shuffle directly to external storage.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 9 above)
Comment (maheshk114) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:414]: @dongjoon-hyun Do you think it can be  modified this way ?

**Comment by attilapiros** [comment]
> > I am also concerned about the performance.
> > I think the best would be if the migration of shuffle data to external storage would only kick in when the scale down is aggressive. This can be decided by checking the ratio of the number of available peers (non-decommissioning executors) and the number of decommissioning executors. In that case the parameter would not be a single boolean flag but a threshold for the ratio.
> > @maheshk114 WDYT?
> 
> Its not only performance but also useful when the nodes are not very reliable. So I think we should have a Boolean flag also to allow user to chose to migrate the shuffle directly to external storage.

With a threshold you can force to do the migration to the external storage every time (or even never) so the flag won't be needed.

**Comment by maheshk114** [comment]
> I am also concerned about the performance.
> 
> I think the best would be if the migration of shuffle data to external storage would only kick in when the scale down is aggressive. This can be decided by checking the ratio of the number of available peers (non-decommissioning executors) and the number of decommissioning executors. In that case the parameter would not be a single boolean flag but a threshold for the ratio.

Yes. Will modify accordingly.

**Comment by maheshk114** [comment]
> > > I am also concerned about the performance.
> > > I think the best would be if the migration of shuffle data to external storage would only kick in when the scale down is aggressive. This can be decided by checking the ratio of the number of available peers (non-decommissioning executors) and the number of decommissioning executors. In that case the parameter would not be a single boolean flag but a threshold for the ratio.
> > > @maheshk114 WDYT?
> > 
> > 
> > Its not only performance but also useful when the nodes are not very reliable. So I think we should have a Boolean flag also to allow user to chose to migrate the shuffle directly to external storage.
> 
> With a threshold you can force to do the migration to the external storage every time (or even never) so the flag won't be needed.

@attilapiros This will work for our use case, as we will be setting it either to 0% or 100%. In normal scenarios, it should be set to more than 50%. That means when more than 50% of executors are decommissioning, we can enable migrating to external storage. But as of now Spark decommissions the  executor as and when the executor is found idle for more than idle timeout or for some reason cluster manager wants to decommission the node. So there is no way to find, how many total executors are going to be decommissioned in advance. With this logic, we will start migrating to peer and later when the number reaches the threshold we will start migrating to external storage. This will result into extra movement of shuffle which we want to avoid in the first place.
